draft-trammell-taps-interface.md

```diff
@@ -758,31 +761,37 @@ configure the remote.
+Groups of Preconnections or Connections can be created using the Clone action:
 
 ~~~
+Preconnection : Preconnection.Clone()
+
 Connection := Connection.Clone()
 ~~~
 
-Calling this once yields a group of two Connections: the parent Connection -- whose
-Clone action was called -- and the resulting clone. Calling Clone on any of these two
-Connections adds a third Connection to the group, and so on.
-All Connections in a group are entangled. This means that they automatically share
-all properties: changing a parameter for one of them also changes the parameter
-for all others.
-
-There is only one Protocol Property that is not entangled, i.e., it is a separate
-per-Connection Property for individual Connections in the group: a priority.
-This priority, which can be represented as a non-negative integer or float, expresses
-a desired share of the Connection Group's available network capacity, such that an
-ideal transport system implementation would assign the Connection the capacity
-share P x C/sum_P, where P = priority, C = total available capacity and sum_P = sum
-of all priority values that are used for the Connections in the same Connection Group.
-The priority setting is purely advisory; no guarantees are given.
-
-Connection Groups should be created by cloning as as early as possible in order
-to aid the Transport System in choosing and configuring the right protocols (see
-also {{transport-params}}).
+Calling Clone on a Connection yields a group of two Connections: the parent
+Connection on which Clone was called, and the resulting clone Connection. These
+connections are "entangled" with each other, and become part of a connection
+group. Calling Clone on any of these two connections adds a third Connection to
+the group, and so on. Connections in a connection group share all their
+properties, and changing the properties on one Connection in the group changes
+the property for all others.
+
+Calling Clone on a Preconnection yields a Preconnection with the same
+parameters, which is "entangled" with the parent Preconnection: all the
+Connections created from entangled Preconnections will be entangled as if they
+had been cloned, and will belong to the same connection group. Calling Clone on
+a Preconnection may be taken by the system an implicit sign that protocol stacks
+supporting efficient connection group usage are preferred by the application.
+
+There is only one Protocol Property that is not entangled, i.e., it is a
+separate per-Connection Property for individual Connections in the group:
+niceness. Niceness works as in {{send-niceness}}: when allocating available
+network capacity among Connections in a connection group, sends on connections
+with higher Niceness values will tend to yield to connections with lower
+Niceness values. The priority setting is purely advisory. No guarantees are
+given, and each implementation is free to implement exact capacity allocation as
+it sees fit.
```

Comment:
"tend to yield to connections" is IMO too weak - it doesn't say anything about how the available capacity is really divided; however, when we DO get a multi-streaming protocol, the capacity division can become quite precise. That's why I included the equation for an "ideal transport system implementation". Why not do the same? Let me try:

An ideal transport system implementation would assign the Connection the capacity share (M-N) x C / M, where N is the Connection's Niceness value, M is the maximum Niceness value used by all Connections in the group and C is the total available capacity (this assumes that positive integer or float values are used to represent Niceness).

Reply:
Niceness in send properties is (was originally) specified as a set of discrete classes, as opposed to an inverse numeric priority, but yeah, this is a better way to do it (as long as the "as it sees fit" language stays in). will change.